editor/gui/editor_scene_tabs.cpp

```diff
@@ -0,0 +1,375 @@
+	update_scene_tabs();
+}
+
+void EditorSceneTabs::_update_context_menu() {
+	scene_tabs_context_menu->clear();
+	scene_tabs_context_menu->reset_size();
+
+	scene_tabs_context_menu->add_shortcut(ED_GET_SHORTCUT("editor/new_scene"), EditorNode::FILE_NEW_SCENE);
+	if (scene_tabs->get_hovered_tab() >= 0) {
+		scene_tabs_context_menu->add_shortcut(ED_GET_SHORTCUT("editor/save_scene"), EditorNode::FILE_SAVE_SCENE);
+		scene_tabs_context_menu->add_shortcut(ED_GET_SHORTCUT("editor/save_scene_as"), EditorNode::FILE_SAVE_AS_SCENE);
+	}
+	scene_tabs_context_menu->add_shortcut(ED_GET_SHORTCUT("editor/save_all_scenes"), EditorNode::FILE_SAVE_ALL_SCENES);
+	if (scene_tabs->get_hovered_tab() >= 0) {
+		scene_tabs_context_menu->add_separator();
+		scene_tabs_context_menu->add_item(TTR("Show in FileSystem"), EditorNode::FILE_SHOW_IN_FILESYSTEM);
+		scene_tabs_context_menu->add_item(TTR("Play This Scene"), EditorNode::FILE_RUN_SCENE);
+
+		scene_tabs_context_menu->add_separator();
+		scene_tabs_context_menu->add_shortcut(ED_GET_SHORTCUT("editor/close_scene"), EditorNode::FILE_CLOSE);
+		scene_tabs_context_menu->set_item_text(scene_tabs_context_menu->get_item_index(EditorNode::FILE_CLOSE), TTR("Close Tab"));
+		scene_tabs_context_menu->add_shortcut(ED_GET_SHORTCUT("editor/reopen_closed_scene"), EditorNode::FILE_OPEN_PREV);
+		scene_tabs_context_menu->set_item_text(scene_tabs_context_menu->get_item_index(EditorNode::FILE_OPEN_PREV), TTR("Undo Close Tab"));
+		if (!EditorNode::get_singleton()->has_previous_scenes()) {
+			scene_tabs_context_menu->set_item_disabled(scene_tabs_context_menu->get_item_index(EditorNode::FILE_OPEN_PREV), true);
+		}
+		scene_tabs_context_menu->add_item(TTR("Close Other Tabs"), EditorNode::FILE_CLOSE_OTHERS);
+		scene_tabs_context_menu->add_item(TTR("Close Tabs to the Right"), EditorNode::FILE_CLOSE_RIGHT);
+		scene_tabs_context_menu->add_item(TTR("Close All Tabs"), EditorNode::FILE_CLOSE_ALL);
+	}
+}
+
+// TODO: This REALLY should be done in a better way than replacing all tabs after almost EVERY action.
+void EditorSceneTabs::update_scene_tabs() {
+	tab_preview_panel->hide();
+
+	bool show_rb = EDITOR_GET("interface/scene_tabs/show_script_button");
+
+	if (DisplayServer::get_singleton()->has_feature(DisplayServer::FEATURE_GLOBAL_MENU)) {
+		DisplayServer::get_singleton()->global_menu_clear("_dock");
+	}
+
+	// Get all scene names, which may be ambiguous.
+	Vector<String> disambiguated_scene_names;
+	Vector<String> full_path_names;
+	for (int i = 0; i < EditorNode::get_editor_data().get_edited_scene_count(); i++) {
+		disambiguated_scene_names.append(EditorNode::get_editor_data().get_scene_title(i));
+		full_path_names.append(EditorNode::get_editor_data().get_scene_path(i));
+	}
+
+	EditorNode::disambiguate_filenames(full_path_names, disambiguated_scene_names);
+
+	// Workaround to ignore the tab_changed signal from the first added tab.
+	scene_tabs->disconnect("tab_changed", callable_mp(this, &EditorSceneTabs::_scene_tab_changed));
```

Comment:
`Object.set_block_signals()` could be used instead.

Reply:
Do we resolve this conversation? I assume that you don't want to use @KoBeWi's advice for this case, @YuriSizov ?

Reply:
I don't think we should resolve this because that's still something to look into. But, as I've mentioned below, this is not in the scope for this PR.